articles/biometric-2fa-in-a-django-application-with-typingdna/index.md

```diff
@@ -0,0 +1,1165 @@
+
+        params = {
+
+            "type": user_type,
+
+            "text_id": text_id,
+
+            "custom_field": custom_field
+
+        }
+
+        return requests.get(url, headers=self.headers, params=params)
+
+
+    def hash_text(self, text):
+
+        return hashlib.sha1((text + text[::-1]).encode()).hexdigest()
+
+
+```
+
+TypingDNA requires a user ID to save typing patterns to a user, so we created a `hash_text` method to hash the user’s ID before sending.
+
+## Saving Users Typing Patterns
+
+After building our helper library, we can start saving our users’ typing patterns in our accounts for future matching. Lets first import the helper library class into our Django application and create an instance of our helper class by supplying the `apiKey` and `apiSecret` we saved from our TypingDNA dashboard earlier.
+
+```Python
+
+from . typingdnahelper import TypingDNA
+
+
+tdna = TypingDNA("apiKey", "apiSecret")
+
+```
+Our `views.py` file libraries importation should look like the image below:
+
+![import](/engineering-education/biometric-2fa-in-a-django-application-with-typingdna/import.png)
+
+We want to update our `enroll` view to save the typing patterns we received in our dashboard. Update the `enroll` view with the code below:
+
+
+```python
+
+def enroll(request):
+
+    if request.method=="POST":
+
+        tp = request.POST.get("tp")
+
+        username = request.session["reg_user"]["username"]
+
+        r = tdna.auto(tdna.hash_text(username), tp)
+
+        if r.status_code == 200:
+
+            user=UserProfile.objects.get(user=User.objects.get(username=username))
+
+            user.typingdna_secured=True
+
+            user.save()
+
+            request.session["typingdna_auth"] = True
+
+            messages.add_message(request, messages.INFO,"You have successfully registered TypingDNA 2FA", "success")
+
+            return redirect("App:videos")
+
+        else:
+
+            messages.add_message(request, messages.INFO,r.json()["message"], "danger")
+
+            return redirect("App:enroll")
+
+    return render(request,"enroll.html")
+
+```
+In the code above, we collected the typing pattern enrolled by the user and stored in the `tp` variable. We then requested the `auto` endpoint of the TypingDNA API to verify if the typing pattern provided matches the typing patterns enrolled for that user . We then checked if the status code of the request was successful (status code 200). If this was the case, we would update the user data in our database to indicate they have been enrolled. Then we will mark the current logged-in session as authenticated with `session["typingdna_auth"] = True` then redirect the user to their dashboard. However, if the authentication was unsuccessful (due to an error from TypingDNA) we will prompt the error message to the user and let them retry the enrollment process.
+
+After successfully enrolling to TypingDNA, we should be redirected to a page similar to the one in the image below:
```

Comment:
Change "After successfully enrolling to TypingDNA, ..." to "After successfully enrolling in TypingDNA, ..."